### DISCUSSION TITLE
Failed to build lammps with deepmd using CMake

### DISCUSSION THREAD

**Discussion by NewSigma** [deepmodeling/deepmd-kit#340] [Q&A] [answered]
Hello everyone, It seems that the documentation of deepmd-kit does not describe how to build lammps with deepmd using CMake. I'm trying it but failed to link the executable at end. Here is part of the output of make:

[100%] Building CXX object CMakeFiles/lammps.dir/public/home/hwb/Coding/lammps-29Oct20/src/USER-DEEPMD/pppm_dplr.cpp.o
/public/home/hwb/Coding/gcc/gcc-7.5.0/bin/g++  -DFFT_FFTW3 -DLAMMPS_GZIP -DLAMMPS_JPEG -DLAMMPS_MEMALIGN=64 -DLAMMPS_OMP_COMPAT=3 -DLAMMPS_PNG -DLAMMPS_SMALLBIG -Dlammps_EXPORTS -I/public/home/hwb/Coding/lammps-29Oct20/src -I/public/home/hwb/software/tensorflow/include -I/public/home/hwb/software/deepmd-kit/include/deepmd -I/public/home/hwb/Coding/lammps-29Oct20/src/KSPACE -I/public/home/hwb/Coding/lammps-29Oct20/src/USER-DEEPMD -I/public/home/hwb/Coding/lammps-29Oct20/Build/styles -I/public/home/hwb/Coding/lammps-29Oct20/src/STUBS  -O3 -DNDEBUG -fPIC   -fopenmp -std=c++11 -o CMakeFiles/lammps.dir/public/home/hwb/Coding/lammps-29Oct20/src/USER-DEEPMD/pppm_dplr.cpp.o -c /public/home/hwb/Coding/lammps-29Oct20/src/USER-DEEPMD/pppm_dplr.cpp
[100%] Linking CXX shared library liblammps.so
/public/home/hwb/software/cmake/bin/cmake -E cmake_link_script CMakeFiles/lammps.dir/link.txt --verbose=1
/public/home/hwb/Coding/gcc/gcc-7.5.0/bin/g++ -fPIC -O3 -DNDEBUG  -shared -Wl,-soname,liblammps.so.0 -o liblammps.so.0 (lots of .o files ...) -Wl,-rpath,/public/home/hwb/software/deepmd-kit/lib:/public/home/hwb/software/tensorflow/lib: /public/home/hwb/software/deepmd-kit/lib/libdeepmd_op.so /public/home/hwb/software/deepmd-kit/lib/libdeepmd.so /public/home/hwb/software/tensorflow/lib/libtensorflow_cc.so /public/home/hwb/software/tensorflow/lib/libtensorflow_framework.so liblammps_mpi_stubs.a /usr/lib64/libjpeg.so /usr/lib64/libpng.so /usr/lib64/libz.so /usr/lib64/libfftw3.so -lm /public/home/hwb/Coding/gcc/gcc-7.5.0/lib64/libgomp.so -lpthread 
/public/home/hwb/software/cmake/bin/cmake -E cmake_symlink_library liblammps.so.0 liblammps.so.0 liblammps.so
make[2]: Leaving directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
[100%] Built target lammps
make  -f CMakeFiles/lmp.dir/build.make CMakeFiles/lmp.dir/depend
make[2]: Entering directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
cd /public/home/hwb/Coding/lammps-29Oct20/Build && /public/home/hwb/software/cmake/bin/cmake -E cmake_depends "Unix Makefiles" /public/home/hwb/Coding/lammps-29Oct20/cmake /public/home/hwb/Coding/lammps-29Oct20/cmake /public/home/hwb/Coding/lammps-29Oct20/Build /public/home/hwb/Coding/lammps-29Oct20/Build /public/home/hwb/Coding/lammps-29Oct20/Build/CMakeFiles/lmp.dir/DependInfo.cmake --color=
make[2]: Leaving directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
make  -f CMakeFiles/lmp.dir/build.make CMakeFiles/lmp.dir/build
make[2]: Entering directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
[100%] Building CXX object CMakeFiles/lmp.dir/public/home/hwb/Coding/lammps-29Oct20/src/main.cpp.o
/public/home/hwb/Coding/gcc/gcc-7.5.0/bin/g++  -DLAMMPS_SMALLBIG -I/public/home/hwb/software/tensorflow/include -I/public/home/hwb/software/deepmd-kit/include/deepmd -I/public/home/hwb/Coding/lammps-29Oct20/src -I/public/home/hwb/Coding/lammps-29Oct20/src/STUBS -I/public/home/hwb/Coding/lammps-29Oct20/Build/includes  -O3 -DNDEBUG -fPIE   -std=c++11 -o CMakeFiles/lmp.dir/public/home/hwb/Coding/lammps-29Oct20/src/main.cpp.o -c /public/home/hwb/Coding/lammps-29Oct20/src/main.cpp
[100%] Linking CXX executable lmp
/public/home/hwb/software/cmake/bin/cmake -E cmake_link_script CMakeFiles/lmp.dir/link.txt --verbose=1
/public/home/hwb/Coding/gcc/gcc-7.5.0/bin/g++  -O3 -DNDEBUG   CMakeFiles/lmp.dir/public/home/hwb/Coding/lammps-29Oct20/src/main.cpp.o  -o lmp  -Wl,-rpath,/public/home/hwb/software/deepmd-kit/lib:/public/home/hwb/software/tensorflow/lib:/public/home/hwb/Coding/lammps-29Oct20/Build: /public/home/hwb/software/deepmd-kit/lib/libdeepmd_op.so /public/home/hwb/software/deepmd-kit/lib/libdeepmd.so /public/home/hwb/software/tensorflow/lib/libtensorflow_cc.so /public/home/hwb/software/tensorflow/lib/libtensorflow_framework.so liblammps.so.0 
liblammps.so.0: undefined reference to `NNPInterModelDevi::compute_std_f(std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > > const&)'
liblammps.so.0: undefined reference to `NNPInter::compute(double&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, std::vector<float, std::allocator<float> > const&, int, LammpsNeighborList const&, int const&, std::vector<float, std::allocator<float> > const&, std::vector<float, std::allocator<float> > const&)'
liblammps.so.0: undefined reference to `NNPInter::compute(double&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, std::vector<float, std::allocator<float> > const&, int, LammpsNeighborList const&, int const&, std::vector<float, std::allocator<float> > const&, std::vector<float, std::allocator<float> > const&)'
liblammps.so.0: undefined reference to `NNPInterModelDevi::compute_std_e(std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > > const&)'
liblammps.so.0: undefined reference to `NNPInterModelDevi::compute(std::vector<double, std::allocator<double> >&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > >&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > >&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > >&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > >&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, std::vector<float, std::allocator<float> > const&, int, LammpsNeighborList const&, int const&, std::vector<float, std::allocator<float> > const&, std::vector<float, std::allocator<float> > const&)'
liblammps.so.0: undefined reference to `select_by_type(std::vector<int, std::allocator<int> >&, std::vector<int, std::allocator<int> >&, int&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, int const&, std::vector<int, std::allocator<int> > const&)'
liblammps.so.0: undefined reference to `NNPInterModelDevi::compute_avg(std::vector<float, std::allocator<float> >&, std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > > const&)'
liblammps.so.0: undefined reference to `NNPInterModelDevi::compute_relative_std_f(std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, float)'
liblammps.so.0: undefined reference to `DeepTensor::compute(std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, std::vector<float, std::allocator<float> > const&, int, LammpsNeighborList const&)'
liblammps.so.0: undefined reference to `DataModifier::compute(std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> >&, std::vector<float, std::allocator<float> > const&, std::vector<int, std::allocator<int> > const&, std::vector<float, std::allocator<float> > const&, std::vector<std::pair<int, int>, std::allocator<std::pair<int, int> > > const&, std::vector<float, std::allocator<float> > const&, int, LammpsNeighborList const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [lmp] Error 1
make[2]: Leaving directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
make[1]: *** [CMakeFiles/lmp.dir/all] Error 2
make[1]: Leaving directory `/public/home/hwb/Coding/lammps-29Oct20/Build'
make: *** [all] Error 2

I modified the CMakeLists.txt of lammps as follows:

set(STANDARD_PACKAGES ASPHERE BODY CLASS2 COLLOID COMPRESS DIPOLE
  GRANULAR KSPACE LATTE MANYBODY MC MESSAGE MISC MLIAP MOLECULE PERI POEMS
  QEQ REPLICA RIGID SHOCK SPIN SNAP SRD KIM PYTHON MSCG MPIIO VORONOI
  USER-ATC USER-AWPMD USER-BOCS USER-CGDNA USER-MESODPD USER-CGSDK USER-COLVARS
  USER-DIFFRACTION USER-DPD USER-DRUDE USER-EFF USER-FEP USER-H5MD USER-LB
  USER-MANIFOLD USER-MEAMC USER-MESONT USER-MGPT USER-MISC USER-MOFFF USER-MOLFILE
  USER-NETCDF USER-PHONON USER-PLUMED USER-PTM USER-QTB USER-REACTION
  USER-REAXC USER-SCAFACOS USER-SDPD USER-SMD USER-SMTBQ USER-SPH USER-TALLY
  USER-UEF USER-VTK USER-QUIP USER-QMMM USER-YAFF USER-ADIOS USER-DEEPMD) #Editted

set(SUFFIX_PACKAGES CORESHELL USER-OMP KOKKOS OPT USER-INTEL GPU)
foreach(PKG ${STANDARD_PACKAGES} ${SUFFIX_PACKAGES})
  option(PKG_${PKG} "Build ${PKG} Package" OFF)
endforeach()

######################################################
# packages with special compiler needs or external libs
######################################################
target_include_directories(lammps PUBLIC $<BUILD_INTERFACE:${LAMMPS_SOURCE_DIR}>)

if(PKG_USER-DEEPMD)
  include_directories(/public/home/hwb/software/tensorflow/include /public/home/hwb/software/deepmd-kit/include/deepmd) #Editted
target_link_libraries(lammps PRIVATE /public/home/hwb/software/deepmd-kit/lib/libdeepmd_op.so PRIVATE /public/home/hwb/software/deepmd-kit/lib/libdeepmd.so PRIVATE /public/home/hwb/software/tensorflow/lib/libtensorflow_cc.so PRIVATE /public/home/hwb/software/tensorflow/lib/libtensorflow_framework.so) #Editted
  target_link_libraries(lmp PRIVATE /public/home/hwb/software/deepmd-kit/lib/libdeepmd_op.so PRIVATE /public/home/hwb/software/deepmd-kit/lib/libdeepmd.so PRIVATE /public/home/hwb/software/tensorflow/lib/libtensorflow_cc.so PRIVATE /public/home/hwb/software/tensorflow/lib/libtensorflow_framework.so)
endif() #Editted

I enabled package KSPACE and USER-DEEPMD.

Build environment:
CentOS 6
GCC 7.5.0
lammps-29Oct20
deepmd-kit 1.3.1
tensorflow 1.14.0

I wonder why I successfully linked liblammps.so.0 but failed to link lmp. How to solve the problem?

**Comment by njzjz** [comment, marked as answer]
https://github.com/deepmd-kit-recipes/lammps-dp-feedstock/blob/2b328eda12ca31826e51a75936838c917e2b61ef/recipe/build.sh#L14
Please check your cmake command and see if you missed any arguments.

**Reply by NewSigma** [reply]
Thank you for your comment. I have modified the cmake command but got the same error. I'm using the following command:

>cmake -D BUILD_LIB=on -D BUILD_SHARED_LIBS=on -DCMAKE_INSTALL_LIBDIR=lib -DPKG_KSPACE=ON -DPKG_USER-DEEPMD=ON -D CMAKE_CXX_FLAGS="-I/public/home/hwb/software/deepmd-kit/include/deepmd -I/public/home/hwb/software/tensorflow/include -L/public/home/hwb/software/deepmd-kit/lib -L/public/home/hwb/software/tensorflow/lib -Wl,--no-as-needed -lrt -ldeepmd_op -ldeepmd -ltensorflow_cc -ltensorflow_framework -Wl,-rpath=/public/home/hwb/software/deepmd-kit/lib -Wl,-rpath=/public/home/hwb/software/tensorflow/lib" ../cmake

Is there something wrong with it?

**Comment by NewSigma** [comment]
It's solved. I appended -DHIGH_PREC and successfully linked. It seems that my deepmd-kit is built using high precision but lammps not. Thank you very much.
